lib/puppet/util/windows/taskscheduler2.rb

```diff
@@ -0,0 +1,707 @@
+  TASK_TRIGGER_SESSION_STATE_CHANGE  = 11
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa382538%28v=vs.85%29.aspx
+  TASK_VALIDATE_ONLY                 = 0x1
+  TASK_CREATE                        = 0x2
+  TASK_UPDATE                        = 0x4
+  TASK_CREATE_OR_UPDATE              = 0x6
+  TASK_DISABLE                       = 0x8
+  TASK_DONT_ADD_PRINCIPAL_ACE        = 0x10
+  TASK_IGNORE_REGISTRATION_TRIGGERS  = 0x20
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa383566(v=vs.85).aspx
+  TASK_LOGON_NONE                           = 0
+  TASK_LOGON_PASSWORD                       = 1
+  TASK_LOGON_S4U                            = 2
+  TASK_LOGON_INTERACTIVE_TOKEN              = 3
+  TASK_LOGON_GROUP                          = 4
+  TASK_LOGON_SERVICE_ACCOUNT                = 5
+  TASK_LOGON_INTERACTIVE_TOKEN_OR_PASSWORD  = 6
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa380747(v=vs.85).aspx
+  TASK_RUNLEVEL_LUA     = 0
+  TASK_RUNLEVEL_HIGHEST = 1
+
+  def initialize(task_name = nil)
+    @task_service = nil
+    @pITask = nil
+
+    # # TODO Do we need COM for OLE?
+    # if ! self.class.com_initialized
+    #   Puppet::Util::Windows::COM.InitializeCom()
+    #   self.class.com_initialized = true
+    # end
+
+    @task_service = WIN32OLE.new('Schedule.Service')
+    @task_service.connect()
+
+    activate(task_name) unless task_name.nil?
+  end
+
+    # TODO SHOULD BE  tested
+    def get_folder_path_from_task(task_name)
+      path = task_name.rpartition('\\')[0]
+  
+      path.empty? ? ROOT_FOLDER : path
+    end
+  
+    # TODO SHOULD BE  tested
+    def get_task_name_from_task(task_name)
+      task_name.rpartition('\\')[2]
+    end
+    
+  # Returns an array of scheduled task names.
+  # By default EVERYTHING is enumerated
+  # option hash
+  #    include_child_folders: recurses into child folders for tasks. Default true
+  #    include_compatibility: Only include tasks which have any of the specified compatibility levels. Default empty array (everything is permitted)
+  #
+  # DONE
+  def enum_task_names(folder_path = ROOT_FOLDER, options = nil)
+    raise Error.new(_('No current task scheduler. Schedule.Service is NULL.')) if @task_service.nil?
+    raise TypeError unless folder_path.is_a?(String)
+
+    options = {} if options.nil?
+    options[:include_child_folders] = true if options[:include_child_folders].nil?
+    options[:include_compatibility] = [] if options[:include_compatibility].nil?
+
+    array = []
+
+    task_folder = @task_service.GetFolder(folder_path)
+
+    task_folder.GetTasks(TASK_ENUM_HIDDEN).each do |task|
+      included = true
+
+      unless !included || options[:include_compatibility].empty?
+        included = included && options[:include_compatibility].include?(task.Definition.Settings.Compatibility)
+      end
+
+      array << task.Path if included
+    end
+    return array unless options[:include_child_folders]
```

Comment:
I think I'd break apart this method a little bit / refactor... I can show you what I mean later.